File: modules/vision-camera-pose-detection/android/src/main/java/com/swinglink/posedetection/MediaPipePoseDetector.kt
package com.swinglink.posedetection

import android.content.Context
import android.graphics.Bitmap
import android.graphics.ImageFormat
import android.graphics.Matrix
import android.media.Image
import android.util.Log
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import java.nio.ByteBuffer

/**
 * Wrapper around MediaPipe Pose Landmarker for body pose detection.
 *
 * Detects 33 MediaPipe landmarks, maps 33→24 joints matching our
 * app's pose model, and returns a flat list of 72 Doubles:
 * [x, y, confidence] for each joint.
 *
 * Coordinate system:
 *   - MediaPipe returns landmark positions normalized 0-1 relative to image
 *   - Image is rotated before inference so coordinates are in display space
 *   - Y is already top-left origin (no flip needed)
 */
class MediaPipePoseDetector(private val context: Context) {

  private var poseLandmarker: PoseLandmarker? = null
  private var initFailed = false
  private var errorCount = 0L

  /** Whether the model loaded successfully and is ready for inference. */
  val isReady: Boolean get() = poseLandmarker != null && !initFailed

  companion object {
    private const val TAG = "PoseDetection"
    private const val MODEL_ASSET = "pose_landmarker_lite.task"
  }

  /**
   * Maps our 24-joint model to MediaPipe landmark indices.
   * Order matches JOINT_NAMES in pose-normalization.ts.
   *
   * -1 means "neck" = computed as midpoint of left_shoulder (11) and right_shoulder (12).
   */
  private val landmarkMapping = listOf(
    // Original 14 joints (indices 0-13)
    0,   // nose
    -1,  // neck (midpoint of shoulders)
    11,  // left_shoulder
    12,  // right_shoulder
    13,  // left_elbow
    14,  // right_elbow
    15,  // left_wrist
    16,  // right_wrist
    23,  // left_hip
    24,  // right_hip
    25,  // left_knee
    26,  // right_knee
    27,  // left_ankle
    28,  // right_ankle
    // New finger joints (indices 14-19)
    17,  // left_pinky
    18,  // right_pinky
    19,  // left_index
    20,  // right_index
    21,  // left_thumb
    22,  // right_thumb
    // New foot joints (indices 20-23)
    29,  // left_heel
    30,  // right_heel
    31,  // left_foot_index
    32,  // right_foot_index
  )

  init {
    loadModel()
  }

  private fun loadModel() {
    try {
      Log.d(TAG, "Initializing MediaPipe PoseLandmarker")

      val modelBuffer = context.assets.open(MODEL_ASSET).use { inputStream ->
        val bytes = inputStream.readBytes()
        ByteBuffer.allocateDirect(bytes.size).also {
          it.put(bytes)
          it.rewind()
        }
      }
      Log.d(TAG, "Model loaded into buffer: ${modelBuffer.capacity()} bytes")

      val baseOptions = BaseOptions.builder()
        .setModelAssetBuffer(modelBuffer)
        .build()

      val options = PoseLandmarker.PoseLandmarkerOptions.builder()
        .setBaseOptions(baseOptions)
        .setRunningMode(RunningMode.IMAGE)
        .setNumPoses(1)
        .setMinPoseDetectionConfidence(0.5f)
        .setMinPosePresenceConfidence(0.5f)
        .setMinTrackingConfidence(0.5f)
        .build()

      poseLandmarker = PoseLandmarker.createFromOptions(context, options)
      Log.d(TAG, "MediaPipe PoseLandmarker created successfully")
    } catch (e: Throwable) {
      Log.e(TAG, "Failed to create PoseLandmarker: ${e.javaClass.simpleName}: ${e.message}")
      initFailed = true
    }
  }

  /**
   * Runs pose detection on pre-copied YUV frame data.
   *
   * Called from a background thread with YUV bytes copied from the camera
   * frame. Converts YUV→RGB, rotates the bitmap, and runs inference.
   * Coordinates are returned in the rotated (display) space.
   */
  fun detectPoseFromYuv(
    yBytes: ByteArray, uBytes: ByteArray, vBytes: ByteArray,
    width: Int, height: Int,
    yRowStride: Int, uvRowStride: Int, uvPixelStride: Int,
    rotationDegrees: Int,
  ): List<Double>? {
    val landmarker = poseLandmarker ?: return null
    if (initFailed) return null

    // YUV → ARGB_8888 Bitmap (downsampled 3x to ~640x360 — reduces allocations ~9x)
    var bitmap = try {
      yuvToRgbBitmap(yBytes, uBytes, vBytes, width, height, yRowStride, uvRowStride, uvPixelStride, downsample = 3)
    } catch (e: Exception) {
      if (errorCount++ % 60L == 0L) {
        Log.e(TAG, "YUV conversion failed (error #$errorCount): ${e.message}")
      }
      return null
    } ?: return null

    // Rotate bitmap so coordinates are in display (portrait) space
    if (rotationDegrees != 0) {
      val matrix = Matrix()
      matrix.postRotate(rotationDegrees.toFloat())
      bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }

    val mpImage = BitmapImageBuilder(bitmap).build()

    val result = try {
      landmarker.detect(mpImage)
    } catch (e: Exception) {
      if (errorCount++ % 60L == 0L) {
        Log.e(TAG, "PoseLandmarker.detect() failed (error #$errorCount): ${e.message}")
      }
      return null
    }

    val landmarks = result.landmarks()
    if (landmarks.isEmpty() || landmarks[0].size < 33) {
      return null
    }

    val poseLandmarks = landmarks[0]
    val output = MutableList(72) { 0.0 }

    for ((index, mpIndex) in landmarkMapping.withIndex()) {
      val offset = index * 3

      if (mpIndex == -1) {
        val leftShoulder = poseLandmarks[11]
        val rightShoulder = poseLandmarks[12]

        output[offset] = ((leftShoulder.x() + rightShoulder.x()) / 2.0).toDouble()
        output[offset + 1] = ((leftShoulder.y() + rightShoulder.y()) / 2.0).toDouble()
        val leftVis = leftShoulder.visibility()
        val rightVis = rightShoulder.visibility()
        output[offset + 2] = minOf(
          if (leftVis.isPresent) leftVis.get() else 0f,
          if (rightVis.isPresent) rightVis.get() else 0f
        ).toDouble()
        continue
      }

      if (mpIndex >= poseLandmarks.size) continue
      val lm = poseLandmarks[mpIndex]

      output[offset] = lm.x().toDouble()
      output[offset + 1] = lm.y().toDouble()
      val vis = lm.visibility()
      output[offset + 2] = if (vis.isPresent) vis.get().toDouble() else 0.0
    }

    return output
  }

  /**
   * Convert YUV_420_888 byte arrays directly to an ARGB_8888 Bitmap.
   * Handles variable row/pixel strides across devices.
   */
  private fun yuvToRgbBitmap(
    yBytes: ByteArray, uBytes: ByteArray, vBytes: ByteArray,
    width: Int, height: Int,
    yRowStride: Int, uvRowStride: Int, uvPixelStride: Int,
    downsample: Int = 1,
  ): Bitmap? {
    val outW = width / downsample
    val outH = height / downsample
    val pixels = IntArray(outW * outH)

    for (outRow in 0 until outH) {
      for (outCol in 0 until outW) {
        val srcRow = outRow * downsample
        val srcCol = outCol * downsample
        val yIndex = srcRow * yRowStride + srcCol
        val uvRow = srcRow shr 1
        val uvCol = srcCol shr 1
        val uvIndex = uvRow * uvRowStride + uvCol * uvPixelStride

        val y = (yBytes[yIndex].toInt() and 0xFF).toFloat()
        val u = (uBytes[uvIndex].toInt() and 0xFF).toFloat() - 128f
        val v = (vBytes[uvIndex].toInt() and 0xFF).toFloat() - 128f

        val r = (y + 1.370705f * v).toInt().coerceIn(0, 255)
        val g = (y - 0.337633f * u - 0.698001f * v).toInt().coerceIn(0, 255)
        val b = (y + 1.732446f * u).toInt().coerceIn(0, 255)

        pixels[outRow * outW + outCol] = (0xFF shl 24) or (r shl 16) or (g shl 8) or b
      }
    }

    return Bitmap.createBitmap(pixels, outW, outH, Bitmap.Config.ARGB_8888)
  }
}

File: modules/vision-camera-pose-detection/ios/MediaPipePoseDetector.swift
import Foundation
import UIKit
import MediaPipeTasksVision

/**
 * Wrapper around MediaPipe Pose Landmarker for body pose detection.
 *
 * Detects 33 MediaPipe landmarks from copied pixel data, maps 33→24 joints
 * matching our app's pose model, and returns a flat [Double] array
 * of 72 values: [x, y, confidence] for each joint.
 *
 * Coordinate system:
 *   - MediaPipe returns landmark positions normalized 0-1 relative to image
 *   - Image is physically rotated to display orientation before inference
 *   - A 180° coordinate correction compensates for the CIImage→CGImage pipeline
 *   - Y is in top-left origin (y increases downward)
 */
final class MediaPipePoseDetector {

  private var poseLandmarker: PoseLandmarker?
  private var initFailed = false

  /// Cached CIContext — creating one per frame is very expensive.
  private let ciContext = CIContext()

  /// Whether the model loaded successfully and is ready for inference.
  var isReady: Bool { poseLandmarker != nil && !initFailed }

  /// Maps our 24-joint model to MediaPipe landmark indices.
  /// Order matches JOINT_NAMES in pose-normalization.ts.
  ///
  /// "neck" = midpoint of landmarks 11 (left_shoulder) and 12 (right_shoulder)
  private struct JointMapping {
    let mpIndex: Int?       // nil for computed joints (neck)
    let name: String

    /// For computed joints like "neck", store the two source indices.
    let midpointOf: (Int, Int)?

    init(mpIndex: Int, name: String) {
      self.mpIndex = mpIndex
      self.name = name
      self.midpointOf = nil
    }

    init(midpointOf: (Int, Int), name: String) {
      self.mpIndex = nil
      self.name = name
      self.midpointOf = midpointOf
    }
  }

  private static let jointMappings: [JointMapping] = [
    // Original 14 joints (indices 0-13)
    JointMapping(mpIndex: 0, name: "nose"),
    JointMapping(midpointOf: (11, 12), name: "neck"),
    JointMapping(mpIndex: 11, name: "leftShoulder"),
    JointMapping(mpIndex: 12, name: "rightShoulder"),
    JointMapping(mpIndex: 13, name: "leftElbow"),
    JointMapping(mpIndex: 14, name: "rightElbow"),
    JointMapping(mpIndex: 15, name: "leftWrist"),
    JointMapping(mpIndex: 16, name: "rightWrist"),
    JointMapping(mpIndex: 23, name: "leftHip"),
    JointMapping(mpIndex: 24, name: "rightHip"),
    JointMapping(mpIndex: 25, name: "leftKnee"),
    JointMapping(mpIndex: 26, name: "rightKnee"),
    JointMapping(mpIndex: 27, name: "leftAnkle"),
    JointMapping(mpIndex: 28, name: "rightAnkle"),
    // New finger joints (indices 14-19)
    JointMapping(mpIndex: 17, name: "leftPinky"),
    JointMapping(mpIndex: 18, name: "rightPinky"),
    JointMapping(mpIndex: 19, name: "leftIndex"),
    JointMapping(mpIndex: 20, name: "rightIndex"),
    JointMapping(mpIndex: 21, name: "leftThumb"),
    JointMapping(mpIndex: 22, name: "rightThumb"),
    // New foot joints (indices 20-23)
    JointMapping(mpIndex: 29, name: "leftHeel"),
    JointMapping(mpIndex: 30, name: "rightHeel"),
    JointMapping(mpIndex: 31, name: "leftFootIndex"),
    JointMapping(mpIndex: 32, name: "rightFootIndex"),
  ]

  init() {
    loadModel()
  }

  private func loadModel() {
    // Look for the .task model file in multiple bundle locations:
    // 1. Pod's own bundle (static linking → usually Bundle.main)
    // 2. Named resource bundle (from resource_bundles podspec config)
    // 3. Main app bundle (fallback)
    let podBundle = Bundle(for: MediaPipePoseDetector.self)
    let resourceBundle = podBundle.url(forResource: "VisionCameraPoseDetection", withExtension: "bundle")
      .flatMap { Bundle(url: $0) }
      ?? Bundle.main.url(forResource: "VisionCameraPoseDetection", withExtension: "bundle")
      .flatMap { Bundle(url: $0) }

    guard let modelPath = resourceBundle?.path(forResource: "pose_landmarker_lite", ofType: "task")
            ?? podBundle.path(forResource: "pose_landmarker_lite", ofType: "task")
            ?? Bundle.main.path(forResource: "pose_landmarker_lite", ofType: "task") else {
      NSLog("[PoseDetection] pose_landmarker_lite.task not found — pose detection disabled")
      NSLog("[PoseDetection] Searched pod bundle: \(podBundle.bundlePath)")
      NSLog("[PoseDetection] Searched resource bundle: \(resourceBundle?.bundlePath ?? "nil")")
      NSLog("[PoseDetection] Searched main bundle: \(Bundle.main.bundlePath)")
      initFailed = true
      return
    }

    NSLog("[PoseDetection] Found model at: \(modelPath)")

    do {
      let options = PoseLandmarkerOptions()
      options.baseOptions.modelAssetPath = modelPath
      options.runningMode = .image
      options.numPoses = 1
      options.minPoseDetectionConfidence = 0.5
      options.minPosePresenceConfidence = 0.5
      options.minTrackingConfidence = 0.5

      poseLandmarker = try PoseLandmarker(options: options)
      NSLog("[PoseDetection] MediaPipe PoseLandmarker loaded successfully")
    } catch {
      NSLog("[PoseDetection] Failed to create PoseLandmarker: \(error.localizedDescription)")
      initFailed = true
    }
  }

  /**
   * Runs pose detection on pre-copied pixel data.
   *
   * Called from a background queue with pixel bytes copied from the camera
   * frame. Reconstructs a CVPixelBuffer, physically rotates the image to
   * match the display orientation, then runs inference.
   *
   * We physically rotate the CGImage rather than relying on UIImage
   * orientation metadata because MediaPipe's MPImage may not correctly
   * apply UIImage orientation when extracting landmarks — it can detect
   * the pose but return coordinates in the raw buffer space.
   * This matches Android's approach (Bitmap.createBitmap with Matrix rotation).
   */
  func detectPoseFromPixelData(
    _ pixelData: Data,
    width: Int,
    height: Int,
    bytesPerRow: Int,
    pixelFormat: OSType,
    orientation: UIImage.Orientation
  ) -> [Double]? {
    guard let landmarker = poseLandmarker, !initFailed else {
      return nil
    }

    // Reconstruct a CVPixelBuffer from the copied bytes
    var optionalBuffer: CVPixelBuffer?
    let status = CVPixelBufferCreate(
      kCFAllocatorDefault,
      width, height,
      pixelFormat,
      nil,
      &optionalBuffer
    )
    guard status == kCVReturnSuccess, let newBuffer = optionalBuffer else {
      return nil
    }

    CVPixelBufferLockBaseAddress(newBuffer, [])
    if let dest = CVPixelBufferGetBaseAddress(newBuffer) {
      let destBytesPerRow = CVPixelBufferGetBytesPerRow(newBuffer)
      // Copy row by row to handle potentially different bytesPerRow
      pixelData.withUnsafeBytes { srcPtr in
        guard let srcBase = srcPtr.baseAddress else { return }
        let copyWidth = min(bytesPerRow, destBytesPerRow)
        for row in 0..<height {
          let srcRow = srcBase.advanced(by: row * bytesPerRow)
          let dstRow = dest.advanced(by: row * destBytesPerRow)
          memcpy(dstRow, srcRow, copyWidth)
        }
      }
    }
    CVPixelBufferUnlockBaseAddress(newBuffer, [])

    // Convert pixel buffer to CGImage
    let ciImage = CIImage(cvPixelBuffer: newBuffer)
    guard let rawCGImage = ciContext.createCGImage(ciImage, from: ciImage.extent) else {
      return nil
    }

    // Physically rotate the image to display orientation so MediaPipe gets
    // an upright person (required for reliable detection).
    // UIGraphicsImageRenderer handles the UIImage orientation transform.
    let rawUIImage = UIImage(cgImage: rawCGImage, scale: 1.0, orientation: orientation)
    let targetSize = rawUIImage.size  // .size accounts for orientation (swaps w/h for 90° rotations)
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1.0  // Avoid screen-scale multiplication
    let renderer = UIGraphicsImageRenderer(size: targetSize, format: format)
    let orientedUIImage = renderer.image { _ in
      rawUIImage.draw(in: CGRect(origin: .zero, size: targetSize))
    }

    let uiImage = orientedUIImage

    guard let mpImage = try? MPImage(uiImage: uiImage) else {
      return nil
    }

    // Run inference
    guard let result = try? landmarker.detect(image: mpImage),
          let landmarks = result.landmarks.first,
          landmarks.count >= 33 else {
      return nil
    }

    // Map 33 MediaPipe landmarks → 24 joints
    var output = [Double](repeating: 0.0, count: 72)

    for (index, mapping) in Self.jointMappings.enumerated() {
      let offset = index * 3

      if let midpointOf = mapping.midpointOf {
        let (idx1, idx2) = midpointOf
        guard idx1 < landmarks.count, idx2 < landmarks.count else { continue }
        let lm1 = landmarks[idx1]
        let lm2 = landmarks[idx2]

        output[offset] = Double((lm1.x + lm2.x) / 2.0)
        output[offset + 1] = Double((lm1.y + lm2.y) / 2.0)
        output[offset + 2] = Double(min(lm1.visibility?.floatValue ?? 0, lm2.visibility?.floatValue ?? 0))
      } else if let mpIndex = mapping.mpIndex {
        guard mpIndex < landmarks.count else { continue }
        let lm = landmarks[mpIndex]

        output[offset] = Double(lm.x)
        output[offset + 1] = Double(lm.y)
        output[offset + 2] = Double(lm.visibility?.floatValue ?? 0)
      }
    }

    // Correct for 180° offset in the CIImage → CGImage → UIKit rotation pipeline.
    // Both CGContext and UIGraphicsImageRenderer rotation approaches produce a
    // consistent 180° error (the skeleton maps correctly to joints but is upside
    // down). This is likely from CIImage's bottom-left coordinate origin causing
    // a y-flip in the CGImage, which compounds with the rotation to produce 180°.
    // Flipping both coordinates cancels this out.
    for i in stride(from: 0, to: output.count, by: 3) {
      output[i] = 1.0 - output[i]
      output[i + 1] = 1.0 - output[i + 1]
    }

    return output
  }

}
